File: MatlabBinaryFiles/readAISData.m
function [data, error] = readAISData(fid, fileInfo, data)
% reads binary data stored by the AIS Processing module.
%
% Inputs:
%   fid = file identifier
%   fileInfo = structure holding the file header, module header, a handle
%   data = a structure containing the standard data
%
% Output:
%   data = structure containing data from a single object
%

% initialize variables
error=false;

try
    
    % read AIS Processing Module specific data
    dataLength = fread(fid, 1, 'int32');
    if (dataLength==0)
        return;
    end

    data.mmsiNumber = fread(fid, 1, 'int32');
    data.fillBits = fread(fid, 1, 'int16');
    [strVal, strLen] = readJavaUTFString(fid);
    data.charData = strVal;
    [strVal, strLen] = readJavaUTFString(fid);
    data.aisChannel = strVal;
    
    
catch mError
    disp('Error reading AIS data object.  Data read:');
    disp(data);
    disp(getReport(mError));
    error=true;
end

File: R/readAISData.R
#' @title Read AIS Data
#' 
#' @description Reads binary data stored by the AIS Processing module.
#'   
#' @param fid binary file identifier
#' @param fileInfo structure holding the file header and module header
#' @param data a structure containing standard data
#' @param debug logical flag to show more info on errors
#' @param \dots Arguments passed to other functions
#' 
#' @return a structure containing data from a single object, and a logical
#'   flag if an error has occurred
#' 
#' @author Taiki Sakai \email{taiki.sakai@@noaa.gov}
#' 
readAISData <- function(fid, fileInfo, data, debug=FALSE, ...) {
    error <- FALSE
    
    tryCatch({
        # Read AIS Processing Module specific data
        dataLength <- pamBinRead(fid, 'int32', n=1)
        
        if(dataLength==0) {
            return(list(data=data, error=error))
        }
        
        data$mmsiNumber <- pamBinRead(fid, 'int32', n=1)
        data$fillBits <-pamBinRead(fid, 'int16', n=1)
        data$charData <- readJavaUTFString(fid)$str
        data$aisChannel <- readJavaUTFString(fid)$str
        return(list(data=data, error=error))
    # }, warning = function(w) {
    #     print(paste('Warning occurred: ', w))
    #     return(list(data=data, error=error))
    }, error = function(e) {
        if(debug) {
            print(paste0('Error reading ', fileInfo$fileHeader$moduleType, ' Data read:'))
            print(data)
            print(e)
        }
        error <- TRUE
        return(list(data=data, error=error))
    })
}
